ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/target/VirtualArtifactRepository.java

```diff
@@ -0,0 +1,215 @@
+				throw new FileNotFoundException(bundleInfo.getSymbolicName());
+			}
+			return location.toURL().openStream();
+		} else if (artifactModel instanceof IFeatureModel featureModel) {
+			String installLocation = featureModel.getInstallLocation();
+			if (installLocation != null) {
+				File featureJar = new File(installLocation);
+				if (featureJar.isFile()) {
+					return new FileInputStream(featureJar);
+				}
+			}
+			IFeature feature = featureModel.getFeature();
+			// Generate in-memory feature.xml
+			ByteArrayOutputStream featureContent = new ByteArrayOutputStream();
+			try (PrintWriter writer = new PrintWriter(featureContent)) {
+				feature.write("", writer); //$NON-NLS-1$
+			}
+			// Generate in-memory feature jar (with only the feature.xml)
+			try (Jar jar = new Jar(feature.getId() + '_' + feature.getVersion())) {
+				jar.putResource("feature.xml", new WriteResource() { //$NON-NLS-1$
+					@Override
+					public void write(OutputStream out) throws Exception {
+						out.write(featureContent.toByteArray());
+					}
+
+					@Override
+					public long lastModified() {
+						return 0L;
+					}
+				});
+				ByteArrayOutputStream featureJar = new ByteArrayOutputStream();
+				jar.write(featureJar);
+				return new ByteArrayInputStream(featureJar.toByteArray());
+			}
+		}
+		throw new IllegalArgumentException(artifactModel.toString());
+	}
+
+	@Override
+	public IQueryable<IArtifactDescriptor> descriptorQueryable() {
+		return (query, monitor) -> query.perform(artifacts.keySet().iterator());
+	}
+
+	@Override
+	public IQueryResult<IArtifactKey> query(IQuery<IArtifactKey> query, IProgressMonitor monitor) {
+		return new QueryResult<>(artifacts.keySet().stream().map(IArtifactDescriptor::getArtifactKey).iterator());
+	}
+
+	@Override
+	public boolean contains(IArtifactDescriptor descriptor) {
+		return artifacts.containsKey(descriptor);
+	}
+
+	@Override
+	public boolean contains(IArtifactKey key) {
+		return artifacts.keySet().stream().anyMatch(descriptor -> key.equals(descriptor.getArtifactKey()));
+	}
+
+	@Override
+	public IStatus getArtifact(IArtifactDescriptor descriptor, OutputStream destination, IProgressMonitor monitor) {
+		return getRawArtifact(descriptor, destination, monitor);
+	}
+
+	@Override
+	public IArtifactDescriptor[] getArtifactDescriptors(IArtifactKey key) {
+		return artifacts.keySet().stream() //
+				.filter(descriptor -> key.equals(descriptor.getArtifactKey())) //
+				.toArray(IArtifactDescriptor[]::new);
+	}
+
+	@Override
+	public IStatus getArtifacts(IArtifactRequest[] requests, IProgressMonitor monitor) {
+		MultiStatus multiStatus = new MultiStatus(getClass(), IStatus.INFO, "Perform Artifact Requests"); //$NON-NLS-1$
+		SubMonitor subMonitor = SubMonitor.convert(monitor, requests.length);
+		boolean isOk = true;
+		for (IArtifactRequest request : requests) {
+			request.perform(this, subMonitor.split(1));
+			multiStatus.add(request.getResult());
+			isOk &= request.getResult().isOK();
```

Comment:
`MultiStatus` already accumulates the severity of all children as its own severity.
So you can just check below if `multiStatus.isOK()` in the end or just return the `multiStatus` as it is without the need collect its OK-status by yourself.

Reply:
I think this is a common pattern in eclipse/equinox because the multi status can contain further messages. Also the Multistatus usually contains a message that indicates the error / action that failed so it could eb confusing to get a status that is okay but has a message that indicates an error...